homeassistant/helpers/entity.py

```diff
@@ -425,6 +428,20 @@ def _name_translation_key(self) -> str | None:
             f".{self.translation_key}.name"
         )
 
+    def _name_placeholders(self, name: str) -> str:
+        """Process possible placeholders in entity name."""
+        tuples = string.Formatter().parse(name)
+        if len(tuple(tuples)) == 1:
+            return name
+        if (translation_placeholders := self.translation_placeholders) is not None:
+            if TYPE_CHECKING:
+                assert isinstance(translation_placeholders, Mapping)
+            try:
+                return name.format(**translation_placeholders)
+            except KeyError as err:
+                raise HomeAssistantError("Missing placeholder %s" % err) from err
```

Comment:
How will we handle if someone in lokalize accidentaly uses the wrong placeholder. I have seen this before (not intentional copy paste errors). Lokalize does notify the user that its wrong, but doesn't enforce it and we just download them normally.

Reply:
Could we maybe avoid this by having a check in the translation downloader part in the pipeline that checks if every placeholder field is present in the other languages? (and default to english if there's something wrong?)

Reply:
> How will we handle if someone in lokalize accidentaly uses the wrong placeholder. I have seen this before (not intentional copy paste errors). Lokalize does notify the user that its wrong, but doesn't enforce it and we just download them normally.

That is a lokalise issue, not sure how that is related to this PR?

Reply:
the fact that this is still a risk where there's a translation without matching slot

Reply:
But this PR doesn't change that? Like that is already existing, right?

Reply:
that's correct, but we already have this risk with placeholders in for example config entry titles

Reply:
Yes, but in that case the frontend raises an error. What will happen if that same issue happens here.

Reply:
> What will happen if that same issue happens here.

the related entity becomes unavailable and you will find the error messages in the log

Reply:
That is a good point @joostlek 👍

Reply:
i've an idea how to do the fallback in this case (_do the validation of missing/wrong placeholders somewhere in https://github.com/home-assistant/core/blob/512902fc59e1c3b0d7faec4e72745e7b5918e39f/homeassistant/helpers/translation.py#L217_) ... will update the PR later this evening

Reply:
would it be ok to fallback to English at all (_whole `strings.json`_) or do we need to only fallback for the faulty translation strings?

Reply:
I think this validation will become expansive, since every entity name string needs to be parsed twice (_in case any other language then English is used_) by `string.Formatter().parse` (_or maybe some regex_) to figure out if there are placeholders, when yes, if they are equal. Not sure if this validation should be moved into the translation download pipeline 🤔

Reply:
the current implementation (59be6fb) will do the fallback for a category (_eq. 'entity', 'config' or 'title'_), but not the whole `strings.json`

example error message
`2023-11-24 22:27:31.659 ERROR (MainThread) [homeassistant.helpers.translation] Validation of placeholders for localized (de) strings failed. Expected placeholders: {'component.fritzbox.entity.binary_sensor.alarm.name': {'placeholder'}} Gathered placeholders: {'component.fritzbox.entity.binary_sensor.alarm.name': {'placehol'}}`

Reply:
I mean, in my second comment I wanted to propose to keep this check at the place where we download the strings into core. (Need to double check if not translated strings are null or English) either way, we could do this check at that place.

Not sure what the best way to solve this here.

Reply:
The download is done during image [build](https://github.com/home-assistant/core/blob/4860daf1f9a02ecf43d5cb97e16eefd44bd240ff/.github/workflows/builder.yml#L66) with [`script/translations/download.py`](https://github.com/home-assistant/core/blob/dev/script/translations/download.py).

So doing the validation there may delay the delivery of a release, even if there is only one placeholder issue with one out of (_current_) 25191 `strings.json` files. On the other side we would be able to react on this issue during build time and fix it on lokalise.com (_not sure how fast this is possible_)

While doing the validation within code itself would not delay the release, but may let to user issue reports which needs fix on lokalise.com.

Reply:
If we already download the translations within the nightly builds, than we should really move the validation into the download script.

Reply:
Let's not fail, let make sure we use an empty string in case the placeholder is missing.

Reply:
with current one, we do not fail, but fallback to the previous cached resources (_which are the English ones_).
So user might see some English phrases and also the error message in the logs

Reply:
That is, however, not fixing the case.

For example, lets say I have this one: `Wall plug {number}`

And it is translated to: `Stroom plug {nummer}`

Now, `nummer` is a placeholder that isn't accounted for anywhere.

I think the latter additions are to complex of an approach to be honest.

Reply:
in this case `Wall plug {number}` would be used and an additional error message is logged.
Yes, it does not fix it directly, but impact for user is minimal and issue can be solved via an issue report.

Reply:
when thinking about custom components, which comes with own translation files, i think we should keep this fallback code (_our `hassfest` script cannot help in this case_) 🤔 
Nevertheless, we should also do a validation of placeholders in translation during download from lokalise

Reply:
I was discussing this with mib on discord, and I pointed out we have hassfest available as tooling for custom component repos.

In my opinion we should make sure only valid translations flow in, either via checking at download at lokalise (for core) or in hassfest (custom components).

And we probably still need to have some kind of fallback, but that's probably depending on the adoption rate of hassfest.

Reply:
It's not clear to me what has been agreed on and what is already implemented.

A script to find mistakes in translations would be great, but the much more likely issue is that localizations are lagging behind English, and I don't think it's reasonable to block translations because of that.

I also don't see any fallbacks in the code, it appears to me we will fail hard as soon as there is a mismatch between translations and implementation, but maybe I'm missing something?

I think we need to do something like this:
- Try first the local language
- Optionally: If local language fails, try English
- If this fails return name, either without substitution or substituting with empty strings

@mib1185 @joostlek can you confirm please?

Reply:
> I also don't see any fallbacks in the code, it appears to me we will fail hard as soon as there is a mismatch between translations and implementation, but maybe I'm missing something?

we load always the English translation first, then the possible different local language:
https://github.com/home-assistant/core/blob/7975a4605bd637685b10e9b140255ff15b006209/homeassistant/helpers/translation.py#L226

https://github.com/home-assistant/core/blob/7975a4605bd637685b10e9b140255ff15b006209/homeassistant/helpers/translation.py#L236-L242

with this a second iteration in `for component, resource in new_resources.items()` only happens, when a second language is there:
https://github.com/home-assistant/core/blob/7975a4605bd637685b10e9b140255ff15b006209/homeassistant/helpers/translation.py#L318

in the second iteration we now compare the placeholders between both languages
if the local language has the same placeholders, we update the cache with the resources of the local language
https://github.com/home-assistant/core/blob/7975a4605bd637685b10e9b140255ff15b006209/homeassistant/helpers/translation.py#L328-L331

---

tbh it last me a while to understand how this code (_fetching both languages and merge the local language strings with the English ones_) in the original code:
https://github.com/home-assistant/core/blob/c7f6ab2457a9e6f794ae89951c7d787b048bcb8e/homeassistant/helpers/translation.py#L271-L284

Reply:
Yes, the way the local language is loaded is clear. However just because a translation exists for the local language does not mean it's completed. If the local language translation is lagging behind English, the code will currently blow up if the translation placeholders were changed in the English version.

Reply:
I've pushed changes to:
- Don't blow up in `entity.py` if the translations are broken, instead return the name without substituting + log an error
- Don't discard all localized translations if a single translation string has placeholders not matching English, instead just drop the translations with mismatches

IMO, this is now OK, although a lokalize script would be welcome as a separate improvement.